Jetpack Mobile: Scan - ThreatItem

ashiagr opened the pull request:
Parent: #13326

This PR builds threat item state with header + sub header based on different threat types and click listener that will take it to the threat details screen later on.

**To test** 

Prerequisite: Make sure both Scan feature flag and MySiteImprovements flag are set to on in the App settings.

1. Select a site with scan capability on the My Site tab
2. Notice that scan menu is displayed
3. Click scan menu item
4. Notice that threat items are displayed with header and sub header based on threat types as described in the below table:
(Tests included in `ThreatItemBuilderTest`)

References: Calypso's [header](https://opengrok.a8c.com/source/xref/calypso/client/components/jetpack/threat-item-header/index.tsx?r=43d3beb3) & sub header( [1](https://opengrok.a8c.com/source/xref/calypso/client/components/jetpack/threat-item-subheader/index.tsx?r=e940ccc1#83), [2](https://opengrok.a8c.com/source/xref/calypso/client/components/jetpack/threat-item/utils.ts?r=5a6b257a#39))


Threat Type | Header | SubHeader
-- | -- | --
CoreFileModification | {filename}: infected core file | Vulnerability found in WordPress
Database | Database {rows.size} threats | -
File | {filename}: malicious code pattern | Threat found {signature}
VulnerableExtension.ExtensionType.Plugin | Vulnerable Plugin: {slug} (version {version}) | Vulnerability found in plugin
VulnerableExtension.ExtensionType.Theme | Vulnerable Theme: {slug} (version {version}) | Vulnerability found in theme
Generic | Threat found | Miscellaneous vulnerability

**Note**
1. Ignore all styling
2. Not too much detailing is done for the messages e.g. customising `FileThreatModel`'s `SubHeader` based on threat status (`current`/`fixed`/`ignored`)

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ashiagr committed:
Build threat header + sub header based on threat type in ThreatItemBuilder

ashiagr committed:
Update ThreatItem VH

ashiagr committed:
Introduce ThreatItem click listener

ashiagr committed:
Remove unneeded longId override

ashiagr committed:
Introduce ScanNavigationEvents

ashiagr committed:
Add threat item builder tests

ashiagr committed:
Fix lint

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/scan/builders/ThreatItemBuilder.kt:
```diff
@@ -0,0 +1,93 @@
+                    R.string.threat_item_header_vulnerable_theme,
+                    listOf(UiStringText(slug), UiStringText(version))
+                )
+                ExtensionType.UNKNOWN -> throw IllegalArgumentException(
```
⚠️ I'm not sure this is a case in which we should use Fail Fast principle. AFAIU the data are coming from the server. If the server introduces a new ExtensionType I'm not sure if our app (especially old versions of the app) should start crashing. Introduction of a new ExtensionType is not a breaking change from the API point of view. I'd consider wrapping the exception into "if (BuildConfig.DEBUG)" and log a message in production. Are my assumptions correct? Wdyt?

ashiagr replied:
I filtered out `ExtensionType.UNKNOWN` types in fluxc [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/develop/fluxc/src/main/java/org/wordpress/android/fluxc/model/scan/threat/ThreatMapper.kt#L43) so they're returned as `GenericThreatModel` as we might not know how this unknown type should be rendered. Ideally this condition will not be reached.

As suggested, I'll also add this check and a log for production with an appropriate tag once [this PR ](https://github.com/wordpress-mobile/WordPress-Utils-Android/pull/59) to add `AppLog` tags for the jetpack features in `WordPress-Utils-Android` is merged.

malinajirka replied:
Ohh I see, thanks for the clarification. In that case I don't mind crashing - I'm a fan of fail fast principle. :)

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/scan/builders/ThreatItemBuilder.kt:
```diff
@@ -0,0 +1,93 @@
+            when (threatModel.extension.type) {
+                ExtensionType.PLUGIN -> UiStringRes(R.string.threat_item_sub_header_vulnerable_plugin)
+                ExtensionType.THEME -> UiStringRes(R.string.threat_item_sub_header_vulnerable_theme)
+                ExtensionType.UNKNOWN -> throw IllegalArgumentException(
```
Same as above

ashiagr replied:
Same as above

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/scan/builders/ThreatItemBuilder.kt:
```diff
@@ -0,0 +1,93 @@
+        is GenericThreatModel -> UiStringRes(R.string.threat_item_sub_header_misc_vulnerability)
+    }
+
+    private fun getDisplayFileName(fileName: String?) = fileName?.replace(".*/".toRegex(), "") ?: ""
```
💡Idea: We might want to consider adding a comment with examples. It's a bit tricky to review, since I don't know what the code is supposed to do.

malinajirka replied:
Looking at the code again, it's supposed to remove the whole path except of the file name, I can see that now.

I'm wondering if we need to escape "/", wdyt?

ashiagr replied:
> I'm wondering if we need to escape "/", wdyt?

It seems to work well in the `ThreatItemBuilderTest` tests and also displays the file name correctly in the app so looks like we don't need to escape it.

> We might want to consider adding a comment with examples. It's a bit tricky to review, since I don't know what the code is supposed to do.

Good point. Added in 0a92f68

malinajirka commented on WordPress/src/test/java/org/wordpress/android/ui/jetpack/scan/ThreatItemBuilderTest.kt:
```diff
@@ -0,0 +1,236 @@
+    }
+
+    @Test
+    fun `builds threat sub header correctly VulnerableExtensionThreatModel of Plugin type`() {
```
🔍 Nitpick: Missing "for" -> "correctly for VulnerableExtensionThreatModel"

ashiagr replied:
In c5d9fe2

malinajirka commented on WordPress/src/test/java/org/wordpress/android/ui/jetpack/scan/ThreatItemBuilderTest.kt:
```diff
@@ -0,0 +1,236 @@
+    }
+
+    @Test
+    fun `builds threat sub header correctly VulnerableExtensionThreatModel of Theme type`() {
```
🔍 Nitpick: Missing "for" -> "correctly for VulnerableExtensionThreatModel"

ashiagr replied:
In c5d9fe2

malinajirka commented on WordPress/src/test/java/org/wordpress/android/ui/jetpack/scan/ThreatItemBuilderTest.kt:
```diff
@@ -0,0 +1,236 @@
+    }
+
+    @Test
+    fun `builds threat sub header correctly Plugin GenericThreatModel`() {
```
🔍 Nitpick: Missing "for" -> "correctly for VulnerableExtensionThreatModel"

ashiagr replied:
In c5d9fe2

malinajirka reviewed (commented):
Thanks @ashiagr! Great job on exploring the web implementation. It looks great overall. I left one important comment and a couple of minor comments. 



Some issues I found - not related to this PR, but I thought I'd better mention them so we can add them to backlog, if they are not already there.
Issue 1:
First time I open "Scan" menu item after turning on the flag, the screen content is blank (I can see just AppBar title).

Issue 2:
1. Open scan feature for SiteA 
2. Switch to SiteB 
3. Open scan feature for SiteB -> state for SiteA appears for 1-2 seconds

ashiagr committed:
Add missing "for" in tests description

ashiagr committed:
Add a comment describing the regex to get file name

ashiagr committed:
Add regex example

ashiagr commented:
Thanks for the review @malinajirka! I've addressed your comments and added [these issues](https://github.com/wordpress-mobile/WordPress-Android/pull/13646#pullrequestreview-557101253) in the backlog #13326 under known issues.

Ready for another round 🙇‍♀️.

PS> I'll assign @zwarm  as reviewer tomm., enjoy your 🎄 vacations!

malinajirka reviewed (approved):
LGTM, thanks for the changes @ashiagr  ;)
